src/foam/nanos/auth/AgentUserAuthService.java

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.ContextAwareSupport;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.nanos.NanoService;
+import foam.nanos.session.Session;
+
+import static foam.mlang.MLang.EQ;
+import static foam.mlang.MLang.AND;
+
+public class AgentUserAuthService
+  extends    ContextAwareSupport
+  implements AgentAuthService, NanoService
+{
+  protected DAO userDAO_;
+  protected DAO groupDAO_;
+  protected DAO sessionDAO_;
+  protected DAO agentJunctionDAO_;
+
+  public AgentUserAuthService(X x) {
+    setX(x);
+  }
+
+  @Override
+  public void start() {
+    userDAO_     = (DAO) getX().get("localUserDAO");
+    groupDAO_    = (DAO) getX().get("groupDAO");
+    sessionDAO_  = (DAO) getX().get("sessionDAO");
+    agentJunctionDAO_  = (DAO) getX().get("agentJunctionDAO");
+  }
+
+  /* 
+    Sets the currently logged in user as an "agent" within the context &
+    sets "user" in the context to the passed in user. This allows users to
+    act on behalf of others while retaining information on the user.
+  */
+  public User actAs(X x, User superUser) throws AuthenticationException {
+    User agent = (User) x.get("user");
+    User user = (User) userDAO_.find(superUser.getId());
+
+    if ( agent == null ) {
+      throw new AuthenticationException();
+    }
+
+    if ( user == null ) {
+      throw new AuthenticationException("Super user doesn't exist.");
+    }
+
+    Group group = (Group) groupDAO_.find(user.getGroup());
+
+    if ( group == null && ! group.getEnabled() ) {
+      throw new AuthenticationException("Super user must exist within a group.");
+    }
+
+    /*
+      Finds the AgentUserJunction object to see if user can log in as the passed in user.
+      Source users are permitted to act as target users, not vice versa.
+    */
+    UserUserJunction permissionJunction = (UserUserJunction) agentJunctionDAO_.find(AND(
+      EQ(UserUserJunction.SOURCE_ID, agent.getId()),
+      EQ(UserUserJunction.TARGET_ID, user.getId())
```

Comment:
If what I wrote above is true, these should be switched.

Reply:
I've accounted for this within the user user relationship. Source user is permitted to act as target user. I've updated the comment to clarify the term "source" and "target" in this context.

Reply:
That's not the problem. What's stopping Alice from doing this?

```JavaScript
var bob = ctrl.__subContext.userDAO.find(123);

// When Alice adds bob to her list of superusers, the junction is
// created with `sourceId` set to Alice's id and `targetId`
// becomes Bob's id.
ctrl.user.superUsers.add(bob);

// Alice is able to act as Bob without him ever knowing, which is bad.
ctrl.agentAuth.actAs(bob);
```

The way it should be is the other way around. Bob should have to add Alice as an agent, in which case his id would be the `sourceId` and Alice's would be the `targetId`, and in that case, the target is acting as the source, not the source acting as the target.

Reply:
I don't think the source and targetId matter much in this case. This issue would have to be resolved by a custom authenticator on the agentJunctionDAO which I mentioned in the summary.